transaction-impl/src/main/java/com/example/auction/transaction/impl/TransactionRepository.java

```diff
@@ -0,0 +1,195 @@
+import com.lightbend.lagom.javadsl.persistence.AggregateEventTag;
+import com.lightbend.lagom.javadsl.persistence.ReadSide;
+import com.lightbend.lagom.javadsl.persistence.ReadSideProcessor;
+import com.lightbend.lagom.javadsl.persistence.cassandra.CassandraReadSide;
+import com.lightbend.lagom.javadsl.persistence.cassandra.CassandraSession;
+import org.pcollections.PSequence;
+import org.pcollections.TreePVector;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Arrays;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+import static com.example.core.CompletionStageUtils.accept;
+import static com.example.core.CompletionStageUtils.doAll;
+
+@Singleton
+public class TransactionRepository {
+
+    private final CassandraSession session;
+
+    @Inject
+    public TransactionRepository(CassandraSession session, ReadSide readSide) {
+        this.session = session;
+        readSide.register(TransactionEventProcessor.class);
+    }
+
+    CompletionStage<PaginatedSequence<TransactionSummary>> getTransactionsForUser(
+            UUID userId, TransactionInfoStatus status, int page, int pageSize) {
+        return countTransactionSummary(userId, status)
+                .thenCompose(
+                        count -> {
+                            int offset = page * pageSize;
+                            int limit = (page + 1) * pageSize;
+                            CompletionStage<PSequence<TransactionSummary>> transactions = offset > count ?
+                                    CompletableFuture.completedFuture(TreePVector.empty()) :
+                                    selectTransactionSummary(userId, status, offset, limit);
+                            return transactions.thenApply(seq -> new PaginatedSequence<>(seq, page, pageSize, count));
+                        }
+                );
+    }
+
+    private CompletionStage<Integer> countTransactionSummary(UUID userId, TransactionInfoStatus status) {
+        return session
+                .selectOne(
+                        "SELECT COUNT(*) FROM transactionSummary " +
+                                "WHERE (creatorId = ? OR winnerId = ?) AND status = ? " +
+                                "ORDER BY itemId DESC",
+                        userId,
+                        userId,
+                        status
+                )
+                .thenApply(row -> (int) row.get().getLong("count"));
+    }
+
+    private CompletionStage<PSequence<TransactionSummary>> selectTransactionSummary(
+            UUID userId, TransactionInfoStatus status, long offset, int limit) {
+        return session
+                .selectAll(
+                        "SELECT * FROM transactionSummary " +
+                                "WHERE (creatorId = ? OR winnerId = ?) AND status = ? " +
+                                "ORDER BY itemId DESC " +
```

Comment:
I just had a look through, and it looks like it will be a lot of extra work to add the auction completion date, so maybe we can do that as a later improvement. For now, ordering by `itemId DESC` is OK, and it's likely to be approximately the same as ordering by the auction completion date. The `itemId` is a time-based UUID, so it will increase over time.